### DISCUSSION TITLE
Way to render `<Modal isOpen />` on server side

### DISCUSSION THREAD

**Discussion by QzCurious** [adobe/react-spectrum#6258] [Ideas]
## How it current work

Currently, it's okey to have `<Modal isOpen />` on SSR. But it's only compatible on syntax level. The Modal is open on client side.

https://react-aria-ssr-modal.vercel.app/
https://github.com/QzCurious/react-aria-ssr-modal

https://github.com/adobe/react-spectrum/assets/38932402/811dbf89-939d-4cdc-a527-c869cc094f1e

Since modal use portal under the hood, which is only supported on client side.

## Feature Request

I'm wondering if we could have a `<Modal  noPortal />` flag so that Modal just render at the position where it is. Then `<Modal noPortal isOpen />` is supported during SSR.

**Comment by snowystinger** [comment]
Wouldn't that be problematic on the client? Either it's not going to hydrate correctly if we automatically apply this or it's going to be rendered in a weird/broken place until the prop is changed?

**Reply by QzCurious** [reply]
I'm not sure what is failing hydration and what "automatically" means. Maybe my description is not concise, let me put it in another way:
Radix UI have there [Dialog Portal](https://www.radix-ui.com/primitives/docs/components/dialog#portal) separate from actual content. So as a user, I can render an opened dialog on SSR if I need (primary for progressive enhancement for me). 

https://github.com/adobe/react-spectrum/assets/38932402/665222df-cc59-4365-9c40-f11f510734ed

Since RAC already have its shape on how modal is composited, extract portal out of it might not be ideal. That's why I were suggesting adding a flag for that.

**Reply by snowystinger** [reply]
Ah, I think this would be considered a pretty advanced use-case. You could create your own Modal, something like so https://codesandbox.io/p/sandbox/distracted-orla-jqyr72?file=%2Fsrc%2FApp.js%3A29%2C11

I've mostly just copied from our code, you could probably strip it down further, I stopped when I had a demonstration.

If we added the ability to render outside a portal we'd be in unstable territory because we make assumptions for calculating the position. I'm not sure what other assumptions we've made. We haven't tested anything that is structured like this.
